src/tools/actors/ActorExecutionTrace.java

```diff
@@ -1,84 +1,401 @@
+
+      writeParameter(value, b);
+
+      t.resolvedPromises++;
+    }
+  }
+
+  public static void promiseChained(final long parent, final long child) {
+    if (!VmSettings.ACTOR_TRACING) {
+      return;
+    }
+
+    Thread current = Thread.currentThread();
+
+    if (current instanceof ActorProcessingThread) {
+      ActorProcessingThread t = (ActorProcessingThread) current;
+
+      if (t.getThreadLocalBuffer().remaining() < Events.PromiseChained.size) {
+        swapBuffer(t);
+      }
+
+      ByteBuffer b = t.getThreadLocalBuffer();
+      b.put(Events.PromiseChained.id);
+      b.putLong(parent); // id of the parent
+      b.putLong(child); // id of the chained promise
+      t.resolvedPromises++;
+    }
+  }
+
+  public static void mailboxExecuted(final Mailbox m, final Actor actor) {
+    if (!VmSettings.ACTOR_TRACING) {
+      return;
+    }
+
+    Thread current = Thread.currentThread();
+
+    if (current instanceof ActorProcessingThread) {
+      ActorProcessingThread t = (ActorProcessingThread) current;
+
+      if (t.getThreadLocalBuffer().remaining() < Events.Mailbox.size + 100 * 50) {
+        swapBuffer(t);
+      }
+
+      ByteBuffer b = t.getThreadLocalBuffer();
+      b.put(Events.Mailbox.id);
+      b.putLong(m.getBasemessageId()); //base id for messages
+      b.putLong(actor.getActorId()); //receiver of the messages
+      int idx = 0;
+
+      for (EventualMessage em : m) {
+        if (b.remaining() < (45 + em.getArgs().length * 9)) {
+          swapBuffer(t);
+          b = t.getThreadLocalBuffer();
+          b.put(Events.MailboxContd.id);
+          b.putLong(m.getBasemessageId());
+          b.putLong(actor.getActorId()); //receiver of the messages
+          b.putShort((short) idx);
+        }
+
+        if (em instanceof PromiseSendMessage) {
+          b.put(Events.PromiseMessage.id);
+          b.putLong(((PromiseMessage) em).getPromise().getPromiseId());
+        } else {
+          b.put(Events.BasicMessage.id);
+        }
+
+        b.putLong(em.getSender().getActorId()); // sender
+        b.putLong(em.getCausalMessageId());
+        b.putShort(em.getSelector().getSymbolId());
+        b.putLong(m.getMessageExecutionStart(idx));
+        b.putLong(m.getMessageSendTime(idx));
+
+        Object[] args = em.getArgs();
+        b.put((byte) (args.length - 1)); //num paramaters
+
+        for (int i = 1; i < args.length; i++) {
+          //gonna need a 8 plus 1 byte for most parameter, boolean just use two identifiers.
+          if (args[i] instanceof SFarReference) {
+            Object o = ((SFarReference) args[i]).getValue();
+            writeParameter(o, b);
+          } else {
+            writeParameter(args[i], b);
+          }
+        }
+
+        idx++;
+      }
+    }
+
+    logMemoryUsage();
+  }
+
+  private static void writeParameter(final Object param, final ByteBuffer b) {
+    if (param instanceof SPromise) {
+      b.put((byte) 0x04);
```

Comment:
might be better to use the `enum.id()` (which you might want to add as getter) here.
Also, the age old question in software maintainability: should each if better be a polymorphic method on the class itself that takes the byte buffer? Not sure, but I I would prefer using the enum id as first step, to avoid trivial bugs.